nsqdserver/protocol_v2.go

```diff
@@ -1685,23 +1685,75 @@ func readMPUB(r io.Reader, tmp []byte, topic *nsqd.Topic, maxMessageSize int64,
 		if err != nil {
 			return nil, buffers, protocol.NewFatalClientErr(err, "E_BAD_MESSAGE", "MPUB failed to read message body")
 		}
+
+		//parse ext header or trace
 		msgBody := b.Bytes()[:messageSize]
 
 		traceID := uint64(0)
+		var extJsonBytes []byte
 		var realBody []byte
-		if traceEnable {
+		var extJsonLen uint16
+		if traceEnable && !mpubExt {
 			if messageSize <= nsqd.MsgTraceIDLength {
 				return nil, buffers, protocol.NewFatalClientErr(nil, "E_BAD_MESSAGE",
 					fmt.Sprintf("MPUB invalid message(%d) body size %d for tracing", i, messageSize))
 			}
 			traceID = binary.BigEndian.Uint64(msgBody[:nsqd.MsgTraceIDLength])
 			realBody = msgBody[nsqd.MsgTraceIDLength:]
+		} else if mpubExt {
+			//read two byte header length
+			extJsonLen = binary.BigEndian.Uint16(msgBody[:nsqd.MsgJsonHeaderLength])
+			//check json length, make sure it does not exceed slice length
+			if messageSize <= nsqd.MsgJsonHeaderLength+int32(extJsonLen) {
+				return nil, buffers, protocol.NewFatalClientErr(nil, "E_BAD_BODY",
+					fmt.Sprintf("invalid body size %d in ext json header content length", messageSize))
+			}
+			extJsonBytes = msgBody[nsqd.MsgJsonHeaderLength : nsqd.MsgJsonHeaderLength+extJsonLen]
+			//parse trace id, if there is a json ext
+			if extJsonLen > 0 {
+				jsonHeader, err := simpleJson.NewJson(extJsonBytes)
+				if err != nil {
+					return nil, buffers, protocol.NewClientErr(err, ext.E_INVALID_JSON_HEADER, "fail to parse json header")
+				}
+
+				//parse traceID, if there is any
+				traceIDJson, existInJsonHeader := jsonHeader.CheckGet(ext.TRACE_ID_KEY)
+				if existInJsonHeader {
+					traceIDStr, err := traceIDJson.String()
+					if err != nil {
+						return nil, buffers, protocol.NewClientErr(err, "INVALID_TRACE_ID", "passin trace id should be string")
+					}
+					traceID, err = strconv.ParseUint(traceIDStr, 10, 0)
+					if err != nil {
+						return nil, buffers, protocol.NewClientErr(err, "INVALID_TRACE_ID", "invalid trace id")
+					}
+				}
+				//check compatibility when topic does not support ext
+				if !topic.IsExt() {
+					canIgnoreExt := canIgnoreJsonHeader(topicName, jsonHeader)
+					if allowExtCompatible && canIgnoreExt {
+						nsqd.NsqLogger().Debugf("ext content ignored in topic: %v", topicName)
+						//set json header len to 0
+						extJsonLen = 0
```

Comment:
can not change `extJsonLen=0` here, it will get the wrong msg body in `realBody = msgBody[nsqd.MsgJsonHeaderLength+extJsonLen:]`